File: Graph Theory Peaks/BFS/Q2. All Paths From Source to Target/Q2_AllPathsFromSourcetoTarget.java
class Solution {
    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        
        // List to store all possible paths
        List<List<Integer>> ans = new LinkedList<>();
        
        // List to store current path while doing DFS
        List<Integer> current = new ArrayList<>();
        
        // Start path from node 0
        current.add(0);
        
        // Call DFS from source (0) to destination (last node)
        dfs(0, current, graph, graph.length - 1, ans);
        
        // Return all collected paths
        return ans; 
    }

    private void dfs(int src, List<Integer> current, int graph[][], int dest, List<List<Integer>> ans){
        
        // If we reached destination node
        if(src == dest){
            
            // Add a copy of current path to answer
            ans.add(new ArrayList<>(current));
            return;
        }
        
        // Traverse all neighbors of current node
        for(int n : graph[src]){
            
            // Add neighbor to current path
            current.add(n);
            
            // Recursively explore further
            dfs(n, current, graph, dest, ans);
            
            // Backtrack: remove last added node
            current.remove(current.size() - 1);
        }
    }
}

File: Graph Theory Peaks/BFS/Q2. All Paths From Source to Target/Q2_AllPathsFromSourcetoTarget.py
class Solution(object):
    def allPathsSourceTarget(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: List[List[int]]
        """

        result = []
        
        # Stack will store tuples of (current_node, path_so_far)
        stack = [(0, [0])]
        
        # Target node is the last node
        target = len(graph) - 1
        
        # Continue until stack becomes empty
        while stack:
            cur, route = stack.pop()
            
            # If we reach the target node, store the path
            if cur == target:
                result.append(route)
            else:
                # Explore all neighbors of the current node
                for node in graph[cur]:
                    # Add new path including this neighbor
                    stack.append((node, route + [node]))
        
        return result